homeassistant/components/google_sheets/__init__.py

```diff
@@ -96,12 +95,13 @@ def _append_to_sheet(call: ServiceCall, entry: ConfigEntry) -> None:
 
     async def append_to_sheet(call: ServiceCall) -> None:
         """Append new line of data to a Google Sheets document."""
-
-        entry = cast(
-            ConfigEntry,
-            hass.config_entries.async_get_entry(call.data[DATA_CONFIG_ENTRY]),
+        entry: ConfigEntry | None = hass.config_entries.async_get_entry(
+            call.data[DATA_CONFIG_ENTRY]
         )
-        session: OAuth2Session = hass.data[DOMAIN][entry.entry_id]
+        if not entry:
+            raise ValueError("Invalid config entry: {call.data[DATA_CONFIG_ENTRY]}")
```

Comment:
Why change this to `ValueError`? HomeAssistantError is printed witouth a stack trace upstream, as it's an expected error. ValueError should never bubble up.

Reply:
Our standard is to raise ValueError for bad user input.

Reply:
We raise `vol.Invalid` for bad user input.

Reply:
This is a known error, and so we should raise a known error class. `ValueError` is an unexpected error exception. 

https://github.com/home-assistant/core/blob/dev/homeassistant/components/websocket_api/commands.py#L216-L223

Reply:
I suggest we update the websocket command handler since ValueError is used across the application for bad user input in service handlers.

Reply:
I disagree. Passing a string to `int()` is a bug and raises ValueError. Treating that as an "expected error" is wrong.

Reply:
It's ok to raise `ValueError` if you're used inside a voluptuous schema as it handles it correctly.

We need to be able to distinguish between an expected error (because input validation or invalid state etc) versus something like like a utility function being used incorrectly. The message of a `ValueError` is never user facing and should be presented as such. A HomeAssistantError is our base class, and we know it's ok to show to the user.

```
>>> int("not string")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'not string'
```

Reply:
Raise `vol.Invalid`. Alternative is to update the ConfigEntrySelector to access the `hass` object available in a context and verify the config entry exists.

Reply:
We previously said that `ValueError` is a thing here https://developers.home-assistant.io/docs/integration_quality_scale_index#silver- which I added to that doc after a previous similar discussion on a PR. There are  lot of uses in the code base of `ValueError` for cases like this, not that we can't fix, but just to add some context.

Raising `vol.Invalid`, even outside of voluptuous is something we'd want to encourage? Sounds like the semantics are right, it just seems odd for something outside of voluptuous, and I don't think our answer is to turn every case today that returns `ValueError` into a voluptuous checker.

(IMO both are bad and we need a third thing similar to vol.invalid)

Reply:
I think we should continue the discussion in an architecture discussion and merge this PR for now as is.